Add ability to use 3D physics engine in 2D

rburing opened the pull request:
Adds a new option for the 2D physics engine:

<img width="1493" height="430" alt="image" src="https://github.com/user-attachments/assets/63acee16-a6f5-49b1-b6d0-f9f89ceb228f" />

This allows e.g. using Jolt Physics in 2D. Something like this can be desirable e.g. for stability compared to GodotPhysics2D.

Internally it just locks the bodies to a plane and adds some thickness to the shapes, etc.

Needs some more work, it's still a bit janky ~~and is missing some features like one-way collisions, joints, character bodies~~.

In the current state the [2D Physics Platformer Demo](https://github.com/godotengine/godot-demo-projects/tree/master/2d/physics_platformer) and the [2D Platformer Demo](https://github.com/godotengine/godot-demo-projects/tree/master/2d/platformer) are already playable.

- Closes https://github.com/godotengine/godot-proposals/issues/13313

*This PR is sponsored by My Spare Time™.*

rburing committed:
Add ability to use 3D physics engine in 2D

Just internally add some thickness to the shapes, etc.

rburing committed:
Add one-way collisions to 3D physics server (WIP)

rburing commented:
I added one-way collisions to 3D physics (the concept makes sense there too), so now those are working e.g. in the [2D Physics Platformer Demo](https://github.com/godotengine/godot-demo-projects/tree/master/2d/physics_platformer) (with `RigidBody2D` player) and in the [2D Platformer Demo](https://github.com/godotengine/godot-demo-projects/tree/master/2d/platformer) (with `CharacterBody2D` player).

So far I only implemented one-way collisions in GodotPhysics3D (and not yet in Jolt Physics) since it could be done by adapting the GodotPhysics2D code.

I didn't expose/bind a `one_way_collision` property on 3D shapes here since it isn't necessary to expose it for this feature.

rburing commented:
I added one-way collisions for rigid bodies to Jolt Physics, so now they work e.g. in the 2D Physics Platformer demo.

LeaoLuciano commented:
> I'm currently running into this issue with rigid bodies overlapping when using Jolt Physics in 2D (with this PR):
> Screencast.From.2025-10-17.21-12-40.mp4
> 
> I understand that the penetration will happen without CCD enabled, but I don't understand why it doesn't depenetrate afterwards. This doesn't happen with GodotPhysics2D. I'm not sure how to solve this.
> 
> (Test project: [2D Physics Tests](https://github.com/godotengine/godot-demo-projects/tree/master/2d/physics_tests) > Functional Tests > Box Pyramid.)

I changed ```#define SIZE2TO3(m_size2) (SCALE_2D_TO_3D * Vector3(m_size2.x, m_size2.y, 5))``` to ```#define SIZE2TO3(m_size2) (SCALE_2D_TO_3D * Vector3(m_size2.x, m_size2.y, 100.0))``` and it worked:

https://github.com/user-attachments/assets/ad37432b-0e0e-4e88-8d0e-e7cb1045a97c

rburing committed:
Add one-way collision direction to 3D physics server

rburing commented:
Thanks for that @LeaoLuciano!

I updated the PR. I also added `JoltBody::has_one_way_collision_shapes()` to optimize contact filtering, and I added support for custom one-way collision direction, which allowed to make this PR work without introducing a strange default value in 3D (namely `Vector2.DOWN` is `(0, 1)`, and our embedding is mapping `(x, y)` to `(x, y, 0)`, but `Vector3.DOWN` is `(0, -1, 0)`, so without custom one-way collision direction we would have had to make the default 3D one-way collision direction `Vector3.UP`).

The main thing left to do is one-way collisions for character bodies (as in the 2D platformer demo) in Jolt, probably around `JoltPhysicsDirectSpaceState3D::_body_motion_recover`, `JoltPhysicsDirectSpaceState3D::_cast_motion_impl`, and `JoltPhysicsDirectSpaceState3D::_body_motion_collide`.
